contracts/multisig-prover/src/error.rs

```diff
@@ -41,6 +41,11 @@ pub enum ContractError {
     #[error(transparent)]
     BcsError(#[from] bcs::Error),
 
+    /// Todo, Below error throws: binary operation `==` cannot be applied to type `EncodingError<1024>`
+    /// this is a workaround.
+    #[error("encoding/decoding failure: [0]")]
+    SolEncodingError(String),
```

Comment:
Mapping the original error with `#[from]` is not possible due to the following error:
```
binary operation `==` cannot be applied to type `EncodingError<1024>`
```
So this is the workaround initially i found suitable.

Reply:
There's something i am probably not understanding. If we use `Box<dyn Error>` [here](https://github.com/eigerco/axelar-amplifier/blob/a30bd5645cb78d678c68b46e6f11e46263fbbf99/contracts/multisig-prover/src/encoding/rkyv/mod.rs#L14-L15) like:

```rust
type Result<T> = core::result::Result<T, Box<dyn std::error::Error>>;
```

That's okay for the encoding function helper module. However, then we would need to map such generic error into a [ContractError enum variant](https://github.com/eigerco/axelar-amplifier/blob/a30bd5645cb78d678c68b46e6f11e46263fbbf99/contracts/multisig-prover/src/payload.rs#L56-L57), as the [multisig contract functions are requiring](https://github.com/eigerco/axelar-amplifier/blob/a30bd5645cb78d678c68b46e6f11e46263fbbf99/contracts/multisig-prover/src/encoding/rkyv/mod.rs#L68-L69) .

With the current approach, [we are following what the repos is already doing with other encoding libs](https://github.com/eigerco/axelar-amplifier/blob/a30bd5645cb78d678c68b46e6f11e46263fbbf99/contracts/multisig-prover/src/error.rs#L41-L47) , the only difference is that we cannot use `thiserror` `#[from]` , [so we just pass a string as the message](https://github.com/eigerco/axelar-amplifier/blob/a30bd5645cb78d678c68b46e6f11e46263fbbf99/contracts/multisig-prover/src/encoding/rkyv/mod.rs#L123-L124) (not ideal, i know ..)

Let me know guys, i am open to perform any suggestion, just want to better understand it.

Reply:
I wasn't aware of `#[source]`, thanks ! I think it would be a good idea, but after trying, it seems we are deriving macros [`PartialEq` and `IntoContractError`](https://github.com/eigerco/axelar-amplifier/blob/12d3991763894dec2538ce96d67bc829975f9c88/contracts/multisig-prover/src/error.rs#L6-L7) which seems incompatible with that.